### PR TITLE
Fix google OpenID Connect

### PR DESCRIPTION
**Pull request by mvschaik**
This PR fixes logins with Google OpenID Connect. It merges https://github.com/omab/python-social-auth/pull/687 and https://github.com/omab/python-social-auth/pull/693.

### CODE REVIEW

File: social/backends/open_id.py [hunk 1]
```diff
@@ -360,3 +437,16 @@ def request_access_token(self, *args, **kwargs):
...
+                             headers={'Authorization': 'Bearer {0}'.format(access_token)})
+
+    def get_user_details(self, response):
+        return {
+            'username': response.get('preferred_username'),
+            'email': response.get('email'),
+            'fullname': response.get('name'),
+            'first_name': response.get('given_name'),
+            'last_name': response.get('family_name'),
+        }
```
Comment (nickcatal) [inline social/backends/open_id.py]: This should be closer to

```python
        values = {'username': '', 'email': '', 'fullname': '',
                  'first_name': '', 'last_name': ''}

        fullname = values.get('name') or ''
        first_name = values.get('given_name') or ''
        last_name = values.get('family_name') or ''
        email = values.get('email') or ''

        if not fullname and first_name and last_name:
            fullname = first_name + ' ' + last_name
        elif fullname:
            try:
                first_name, last_name = fullname.rsplit(' ', 1)
            except ValueError:
                last_name = fullname

        username_key = self.setting('USERNAME_KEY') or self.USERNAME_KEY
        values.update({'fullname': fullname, 'first_name': first_name,
                       'last_name': last_name,
                       'username': values.get(username_key) or
                                   (first_name.title() + last_name.title()),
                       'email': email})
        return values
```

File: social/backends/open_id.py [hunk 2]
```diff
@@ -271,17 +275,78 @@ def __init__(self, handle, secret='', issued=0, lifetime=0, assoc_type=''):
...
+    # Override OIDC_ENDPOINT in your subclass to enable autoconfig of OIDC
+    OIDC_ENDPOINT = None
+
+    DEFAULT_SCOPE = ['openid', 'profile', 'email']
     EXTRA_DATA = ['id_token', 'refresh_token', ('sub', 'id')]
-    # Set after access_token is retrieved
-    id_token = None
+    REDIRECT_STATE = False
+    ACCESS_TOKEN_METHOD = 'POST'
+    REVOKE_TOKEN_METHOD = 'GET'
```
Comment (nickcatal) [inline social/backends/open_id.py]: There should be a `USERNAME_KEY = 'preferred_username'`

File: social/backends/open_id.py [hunk 3]
```diff
@@ -350,6 +413,20 @@ def validate_and_return_id_token(self, id_token):
...
             raise AuthTokenError(self, 'Incorrect id_token: nonce')
+
+    def validate_and_return_id_token(self, jws):
+        """
+        Validates the id_token according to the steps at
+        http://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation.
+        """
+        try:
+            # Decode the JWT and raise an error if the sig is invalid
+            id_token = JWS().verify_compact(jws.encode('utf-8'), self.get_jwks_keys())
```
Comment (nickcatal) [inline social/backends/open_id.py]: These long lines should likely be at or below 80 characters to match the rest of the file/app

**Comment by nickcatal** [comment]
Left a few comments that I could see on first glance.

This is what I need to start to flesh out part 2 of #750 so hopefully it gets merged in.

File: social/backends/open_id.py (hunk 1 above)
Comment (mvschaik) [inline social/backends/open_id.py]: I appreciate your suggestion, but these fields have been standardized for OIDC in http://openid.net/specs/openid-connect-core-1_0.html#rfc.section.5.1. I therefore would prefer to use these as a sensible default. I guess most users would override the OpenIdConnectAuth to customize this for the actual values the backend provides (if the backend doesn't adhere to the RFC).

**Comment by mvschaik** [comment]
Thanks for your suggestions, @nickcatal! I've added some fixes. I hope it gets merged in too!

File: social/backends/open_id.py (hunk 1 above)
Comment (nickcatal) [inline social/backends/open_id.py]: I was mostly talking about `username_key = self.setting('USERNAME_KEY') or self.USERNAME_KEY`

Which lets users override the username in their own settings file and exists in other backends.

Comment (mvschaik) [inline social/backends/open_id.py]: I've added the USERNAME_KEY setting.

**Comment by gabejackson** [comment]
Any reason this hasn't been merged yet? Have been working on something similar lately and could lend a hand getting this merged if needed.

**Comment by omab** [comment]
I've ported this PR into [social-core](https://github.com/python-social-auth/social-core/commit/25ca1f51b1e2e22521fdee33ac542ce2bcfb4da0).

Thanks!

**Pull request closed without merging**